internal/controller/engine/engine.go

```diff
@@ -0,0 +1,434 @@
+	// reading watches to start from a buffered channel.
+
+	// Take the write lock. This will block any other callers that want to
+	// update the watches for the same controller.
+	c.mx.Lock()
+	defer c.mx.Unlock()
+
+	// Start new sources.
+	for gvk, w := range start {
+		// The controller's Watch method just calls the StoppableSource's Start
+		// method, passing in its private work queue as an argument. This
+		// will start an informer for the watched kind if there isn't one
+		// running already.
+		//
+		// The watch will stop sending events when either the source is stopped,
+		// or its backing informer is stopped. The controller's work queue will
+		// stop processing events when the controller is stopped.
+		src := NewStoppableSource(e.infs, w.kind, w.handler, w.predicates...)
+		if err := c.ctrl.Watch(src); err != nil {
+			return errors.Wrapf(err, "cannot start watch for %q", gvk)
+		}
+
+		// Record that we're now running this source.
+		c.sources[gvk] = src
+
+		e.log.Debug("Started watching GVK", "controller", name, "watched-gvk", gvk)
+	}
+
+	return nil
+}
+
+// StopWatches stops all watches for the supplied controller, except watches for
+// the GVKs supplied by the keep argument. StopWatches blocks other operations
+// on the same controller if and when it stops a watch. It returns the number of
+// watches that it successfully stopped.
+func (e *ControllerEngine) StopWatches(ctx context.Context, name string, keep ...schema.GroupVersionKind) (int, error) {
+	e.mx.RLock()
+	c, ok := e.controllers[name]
+	e.mx.RUnlock()
+
+	if !ok {
+		return 0, errors.Errorf("controller %q is not running", name)
+	}
+
+	c.mx.RLock()
+	stop := sets.KeySet(c.sources).Difference(sets.New(keep...))
+	c.mx.RUnlock()
+
+	// Don't take the write lock if we actually want to keep all watches.
+	if len(stop) == 0 {
+		return 0, nil
+	}
+
+	c.mx.Lock()
+	defer c.mx.Unlock()
+
+	stopped := 0
+	for gvk := range stop {
+		if err := c.sources[gvk].Stop(ctx); err != nil {
+			return stopped, errors.Wrapf(err, "cannot stop watch for %q", gvk)
+		}
+		delete(c.sources, gvk)
+		e.log.Debug("Stopped watching GVK", "controller", name, "watched-gvk", gvk)
+		stopped++
+	}
+
+	return stopped, nil
+}
+
+// RemoveUnwatchedInformers removes all informers that don't power any
+// controller's watches. This may include informers that weren't started by a
+// watch, e.g. informers that were started by a Get or List call to a cache
+// backed by the same informers as the Engine.
+func (e *ControllerEngine) RemoveUnwatchedInformers(ctx context.Context) error {
+	// Build the set of GVKs any controller is currently watching.
+	e.mx.RLock()
+	watching := make(map[schema.GroupVersionKind]bool, len(e.controllers))
+	for _, c := range e.controllers {
+		c.mx.RLock()
+		for gvk := range c.sources {
+			watching[gvk] = true
+		}
+		c.mx.RUnlock()
+	}
+	e.mx.RUnlock()
+
+	for _, gvk := range e.infs.ActiveInformers() {
```

Comment:
To answer my question, sources are garbage collected per controller only every minute. So chance is high (but not 100%?) that the watch just keeps running.

❗ Also a thought that might matter: when an XRD changes its schema (think of adding a default value), the informer watch (the request) is restarted because the apiserver cuts the connection. But the controller-runtime cache now stays the same, doesn't it? So we don't get another informer store client-side, which means it will lack the effects of the new schema.

Reply:
> So chance is high (but not 100%?) that the watch just keeps running.

Stopping the controller calls `StopWatches` then `RemoveUnwatchedInformers`, so restarting a controller immediately triggers a GC of informers.

> Also a thought that might matter

Based on my controller-runtime/client-go spelunking, I _think_ they handle this internally.

A controller-runtime Informer wraps a client-go SharedIndexInformer. The actual objects are stored in the SharedIndexInformer's index.

I noticed there's a DeltaFIFO DeltaType Replaced. Its docstring says:

> Replaced is emitted when we encountered watch errors and had to do a relist. We don't know if the replaced object has changed.

See https://github.com/kubernetes/client-go/blob/v0.30.0/tools/cache/delta_fifo.go#L170

I can also see that the controller processes Replaced DeltaTypes by updating the object in the backing store:

https://github.com/kubernetes/client-go/blob/v0.30.0/tools/cache/controller.go#L460

My read is:

* The Reflector will call ListAndWatch repeatedly (with backoff)
* ListAndWatch will start a watch that runs until it hits an error
* When the watch hits an error, ListAndWatch falls back to calling list
* list always replaces all listed objects in the store

See:

https://github.com/kubernetes/client-go/blob/v0.30.0/tools/cache/reflector.go#L346